src/session_pool/session_pool.js

```diff
@@ -0,0 +1,148 @@
+import EventEmitter from 'events';
+import _ from 'underscore';
+import { openKeyValueStore } from '../key_value_store';
+import Session from './session';
+
+export const SESSION_POOL_DEFAULTS = {
+    maxPoolSize: 1000,
+    maxSessionAgeSecs: 3000,
+    maxSessionReuseCount: 50,
+
+    persistStateKeyValueStoreId: null,
+    persistStateKey: 'SESSION_POOL_STATE',
+
+    createSessionFunction: null,
+
+};
+
+/**
+ * Handles the session rotation, creation and persistence.
+ *
+ */
+
+// TODO: We should probably add some debug loging
+// TODO: Validation
+export default class SessionPool extends EventEmitter {
+    /**
+     *
+     * @param [options]
+     * @param options.initialPoolSize {Number}
+     * @param options.maxPoolSize {Number}
+     * @param options.maxSessionAgeSecs {Number}
+     * @param options.maxSessionReuseCount {Number}
+     * @param options.persistStateKeyValueStoreId {String}
+     */
+    constructor(options) {
+        super();
+
+        const opts = _.defaults({}, options, SESSION_POOL_DEFAULTS);
+
+        // Pool Configuration
+        this.maxPoolSize = opts.maxPoolSize;
+        this.createSessionFunction = opts.createSessionFunction || this._createSessionFunction;
+
+        // Session configuration
+        // @TODO: Maybe options.sessionOptions / this.sessionOptions?
+        this.maxSessionAgeSecs = opts.maxSessionAgeSecs;
+        this.maxSessionReuseCount = opts.maxSessionReuseCount;
+
+        // Session storage
+        this.persistStateKeyValueStoreId = opts.persistStateKeyValueStoreId;
+        this.persistStateKey = opts.persistStateKey;
+
+        // Statistics
+        this.activeSessions = 0;
+        this.blockedSessions = 0;
+
+        // Operative states
+        this.storage = null;
+        this.sessions = [];
+
+        // Maybe we can add onSessionRetired function to configuration ?
```

Comment:
For that we should have an event, not?

Reply:
Sorry, I don't understand, why having `onSessionRetired` rather then `sessionRetired` event?

Reply:
As discussed with @mnmkng. This listing on `discarded` event would be done only a few times, so having an extra configuration parameter is not necessary.

Reply:
Sure. Anyway, if that's an event, it would be better to call it `sessionDiscarded` rather than just `discarded`, because then you don't really know if that relates to the entire instance or a single session.